docs/design/2020-01-24-collations.md

```diff
@@ -0,0 +1,243 @@
+
+In MySQL, collation is treated as an attribute of the character set: for each character set has options for different collations(and one of them is the default collation), every collation belongs to one character set only. The effects of collation are:
+  * It defines a total order on all characters of a character set.
+  * It determines if padding is applied when comparing two strings.
+  * It describes comparisons of logically identical but physically different byte sequences.
+Some examples of the effects of collation can be found in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-effect.html) of the MySQL Manual and [this webpage](http://demo.icu-project.org/icu-bin/locexp?_=en_US&x=col).
+
+Several collation implementations are described in [this part](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-implementations.html) of MySQL Manual:
+  * Per-character table lookup for 8-bit character sets, for example: `latin1_swedish_ci`.
+  * Per-character table lookup for multibyte character sets, including some of the Unicode collations, for example: `utf8mb4_general_ci`.
+    - Full UCA implementation for some Unicode collations, for example: `utf8mb4_900_ai_ci`.
+
+## Proposal
+
+### Collate Interface
+
+There are two basic functions needed for collation `sortKey` calculation:
+  * Function that is used to compare two strings according to the given collation.
+  * Function that is used to make a memory-comparable `sortKey` corresponding to the given string.
+The interface can be defined as following:
+
+```
+type CollAttrFlag int
+
+const (
+    FlagIgnoreCase uint64=1 << iota
+    FlagIgnoreAccent
+    FlagPadding
+)
+
+type CharsetCollation interface {
+    // Compare returns an integer comparing the two byte slices. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    Compare(a, b []byte) (int, error)
+    // CompareString returns an integer comparing the two strings. The result will be 0 if a==b, -1 if a < b, and +1 if a > b.
+    CompareString(a, b string) (int, error)
+    // Key returns the collation key for str, the returned slice will point to an allocation in Buffer.
+    Key(buf *Buffer, str []byte) error
```

Comment:
the returned slice is in `buf`? I think it is better to return it by multiple return values. like `(buf *Buffer, err error)`

Reply:
Yes, the value is returned in `buf`. If we put the `*Buffer` in return value, it must be allocated in the function, that will be costly.

Reply:
We can pass the buf in, and return it in return values.

Reply:
Like what `go package` does: `func (c *Collator) Key(buf *Buffer, str []byte) []byte
func (c *Collator) KeyFromString(buf *Buffer, str string) []byte`

Reply:
Got it, thanks.